Cabal/doc/nix-local-build.markdown

```diff
@@ -0,0 +1,364 @@
+and look something like this::
+
+~~~~
+constraints: HTTP ==4000.3.3,
+             HTTP +warp-tests -warn-as-error -network23 +network-uri -mtl1 -conduit10,
+             QuickCheck ==2.9.1,
+             QuickCheck +templatehaskell,
+             ...
+~~~~
+
+For end-user applications, you might consider distributing the
+`cabal.project.freeze` file in your source repository so that
+all contributors see a consistent set of dependencies.
+
+# Configuring builds with cabal.project #
+
+`cabal.project` files support a variety of options which configure the
+details of your build.  The general syntax of a `cabal.project` file
+is similar to that of a Cabal file: there are a number of fields, some
+of which live inside stanzas:
+
+~~~~
+packages: */*.cabal
+with-compiler: /opt/ghc/8.0.1/bin/ghc
+
+package cryptohash
+  optimization: False
+~~~~
+
+## Glob syntax ##
+
+Some options in `cabal.project` take globs.  `cabal new-build` has
+its own glob format, which is described by the following BNF:
+
+~~~~
+FilePathGlob    ::= FilePathRoot FilePathGlobRel
+FilePathRoot    ::= {- empty -}        # relative to cabal.project
+                  | "/"                # Unix root
+                  | [a-zA-Z] ":" [/\\] # Windows root
+                  | "~"                # home directory
+FilePathGlobRel ::= Glob "/"  FilePathGlobRel # Unix directory
+                  | Glob "\\" FilePathGlobRel # Windows directory
+                  | Glob         # file
+                  | {- empty -}  # trailing slash
+Glob      ::= GlobPiece *
+GlobPiece ::= "*"            # wildcard
+            | [^*{},/\\] *   # literal string
+            | "\\" [*{},]    # escaped reserved character
+            | "{" Glob "," ... "," Glob "}" # union (match any of these)
+~~~~
+
+Wildcards and unions have the obvious semantics.  A wildcard like
+`*/foo.cabal` does NOT match `.hidden/foo.cabal`; you'll need to
+explicitly write a period to match this.
+
+## Specifying the local packages ##
+
+The following top-level options specify what the local packages of a
+project are:
+
+`packages:` _package location list_ (space separated, default: `./*.cabal`)
+:   Specifies the list of package locations which contain
+    the local packages to be built by this project.  Package locations
+    can take the following forms:
+
+    1. They can specify a Cabal file, or a directory containing
+       a Cabal file, e.g.,
+       `packages: Cabal cabal-install/cabal-install.cabal`
+
+    2. They can specify a glob-style wildcards, which must match one
+       or more (a) directories containing a (single) Cabal file, (b) Cabal
+       files (extension `.cabal`), or (c) ~~tarballs which
+       contain Cabal packages (extension `.tar.gz`)~~ (not implemented yet).  For example, to
+       match all Cabal files in all subdirectories, as well as the Cabal
+       projects in the parent directories `foo` and `bar`, use
+       `packages: */*.cabal ../{foo,bar}/`
```

Comment:
nice, much better examples, could improve the glob syntax section above with examples like this